### PR TITLE
Fix LruCache callback deduplication

### LINKED ISSUE #6200: CacheDecoratorTestCase.test_double_get fails on Python 3.8: 4 != 3
### Description

There is a test failure when synapse 1.4.0 is built with Python 3.8.

### Steps to reproduce

- Run the tests using Python 3.8: `PYTHONPATH=. trial-3 tests`

The test suite fails:
```
[...]
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/usr/lib64/python3.8/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
  File "/builddir/build/BUILD/synapse-1.4.0/tests/storage/test__base.py", line 308, in test_double_get
    self.assertEquals(a.func2.cache.cache.pop.call_count, 3)
  File "/usr/lib64/python3.8/site-packages/twisted/trial/_synctest.py", line 432, in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
  File "/usr/lib64/python3.8/unittest/case.py", line 912, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib64/python3.8/unittest/case.py", line 905, in _baseAssertEqual
    raise self.failureException(msg)
twisted.trial.unittest.FailTest: 4 != 3

tests.storage.test__base.CacheDecoratorTestCase.test_double_get
-------------------------------------------------------------------------------
Ran 780 tests in 194.729s

FAILED (skips=1, failures=1, successes=778)
```

### Version information

- **Version**:
1.4.0

- **Install method**:
running tests from the source tree

- **Platform**:
Fedora rawhide (to become Fedora 32) which includes Python 3.8

### PR DESCRIPTION
**Pull request by V02460**
The PR fixes the deduplication issues for memoizing caches. It makes all identical _CacheContext objects share the same `invalidate` function. This is required for proper deduplication of invalidate callbacks in LruCache. LruCache is used for the `@cached` function decorator. Fixes #6200.

LruCache relies on the equality of callback functions for deduplication. This is in contrast to what was annotated in _CacheContext which claims that LruCache relies on the equality of _CacheContext objects itself.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#changelog)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off)

### CODE REVIEW

**Comment by richvdh** [comment]
yup, so this has been changed by https://github.com/python/cpython/commit/ac20e0f98d6727ba97a9575bfa2a11b2f6247c35#diff-8bd563e1946aef62abb51ecc6d162eaf.

**Comment by V02460** [comment]
Ensuring `_CacheContext` objects are not duplicated seems saner to me as well, right.

I'd argue that it's not CacheDescriptor's responsibility to manage caching of `_CacheContext` instances. Because of this and to not clutter `CacheDescriptor` even more, it might be better to put that code into a `get_instance` factory method of `_CacheContext` instead.

I think we still need to use a `WeakValueDictionary` instead of a normal dict as you suggested. For every new argument of a method call a `_CacheContext` object is created, so these become many and really should be released when no longer used.

**Comment by richvdh** [comment]
> I'd argue that it's not CacheDescriptor's responsibility to manage caching of _CacheContext instances. Because of this and to not clutter CacheDescriptor even more, it might be better to put that code into a get_instance factory method of _CacheContext instead.

ok, that sounds reasonable.

> I think we still need to use a WeakValueDictionary instead of a normal dict as you suggested. For every new argument of a method call a _CacheContext object is created, so these become many and really should be released when no longer used.

I think you're right. I hadn't quite thought that part through!

File: synapse/util/caches/descriptors.py [hunk 1]
```diff
@@ -38,6 +40,8 @@
 
 logger = logging.getLogger(__name__)
 
+CacheKey = Union[Tuple, Any]
```
Comment (richvdh) [inline synapse/util/caches/descriptors.py]: isn't `Union[Tuple, Any]` the same as `Any` ?

Comment (V02460) [reply synapse/util/caches/descriptors.py]: Yes, but it better represents the fact that it is either a single argument or a tuple of arguments.

**Comment by richvdh** [comment]
:tada:

**Comment by richvdh** [comment]
thank you very much for contributing this and for working through the pain from the CI!

**Comment by auscompgeek** [comment]
FYI you didn't have to turn the function annotations into comments - that's valid Python 3 syntax.

**Comment by V02460** [comment]
Oh thank you, I didn't know function annotation syntax was available earlier ([PEP 3107](https://www.python.org/dev/peps/pep-3107), Python 3.0) than variable annotation syntax ([PEP 526](https://www.python.org/dev/peps/pep-0526), Python 3.6).

**Pull request merged**